### PR TITLE
[sweet-api][kotlin] Add better error handling

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Adds better error handling. 

# How

- Added a new promise interface and bridge promise wrapper. This is needed by the new API to automatically handle the `CodedException` and the return type conversion. 
- Added `CodedException` - basic abstraction over all exceptions which should be thrown by the module.
- Added a couple of CodedExpection subclass to handle errors thrown by the `e-m-c` implementation.

# Test Plan

- unit tests

### CODE REVIEW

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/KotlinPromiseWrapper.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    when (value) {
+      is Bundle -> {
+        bridgePromise.resolve(Arguments.fromBundle(value as Bundle?))
+      }
+      is List<*> -> {
+        bridgePromise.resolve(Arguments.fromList(value as List<*>?))
+      }
+      else -> {
+        bridgePromise.resolve(value)
+      }
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/KotlinPromiseWrapper.kt:20]:
```suggestion
      is Bundle -> bridgePromise.resolve(Arguments.fromBundle(value as Bundle?))
      is List<*> -> bridgePromise.resolve(Arguments.fromList(value as List<*>?))
      else -> bridgePromise.resolve(value)
```

File: packages/expo-modules-core/android/src/test/java/expo/modules/kotlin/methods/AnyMethodTest.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    method.call(
+      JavaOnlyArray().apply {
+        pushString("STRING")
+      },
+      promise
+    )
+
+    Truth.assertThat(promise.state).isEqualTo(PromiseState.REJECTED)
+    Truth.assertThat(promise.rejectCode).isEqualTo("ERR_INCOMPATIBLE_ARG_TYPE")
+    Truth.assertThat(promise.rejectMessage).isEqualTo("Type kotlin.String of argument is not compatible with expected type kotlin.Int.")
```
Comment (barthap) [inline packages/expo-modules-core/android/src/test/java/expo/modules/kotlin/methods/AnyMethodTest.kt:68]: Maybe
```suggestion
    Truth.assertThat(promise.rejectMessage).isEqualTo("Argument type kotlin.String is not compatible with expected type kotlin.Int.")
```
That `Type xx of argument...` sounds unnatural 🤔

**Review by barthap — APPROVED** [review]

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/methods/AnyMethod.kt [hunk 3]
```diff
@@ -1,20 +1,41 @@
...
+      callUserImplementation(convertedArgs, promise)
+    } catch (codedError: CodedException) {
+      promise.reject(codedError)
+    } catch (e: Throwable) {
+      promise.reject(UnexpectedException(e))
+    }
+  }
+
+  @Throws(CodedException::class)
+  internal abstract fun callUserImplementation(args: Array<out Any?>, promise: Promise)
```
Comment (bbarthec) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/methods/AnyMethod.kt:33]: `User` meaning is unclear in this context 🤔 
I would suggest getting rid of it or replacing the naming concept altogether.
```suggestion
  internal abstract fun callImplementation(args: Array<out Any?>, promise: Promise)
```
Alternatives: `doCall`, `invoke`.

**Review by bbarthec — COMMENTED** [review]
Overall it looks good to me 👍 
I've got a single remark though:
- I wonder if we can make these errors more crammed together in one file? I'm not sure if that's a good practice or not, but it might ease the pain of searching for an existing exception to reuse it somewhere in the code or create a fresh one basing off the others? My initial though is: exception name (exposed via the file name) is not always sufficient enough to decide between the available ones.

**Review by tsapeta — APPROVED** [review]

**Pull request merged**